File: python/zymeup_sdk/support_ticket.py
"""Zymeup Python SDK - Support Ticket Client"""
from typing import Optional, Dict, Any, List


class SupportTicketClient:
    """Support ticket client"""

    def __init__(self, client):
        self._client = client

    def create(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a support ticket"""
        return self._client.request("POST", "/shipment/support/tickets", json=data)

    def list(self, status: Optional[str] = None) -> Dict[str, Any]:
        """List support tickets"""
        params: Dict[str, Any] = {}
        if status:
            params["status"] = status
        return self._client.request("GET", "/shipment/support/tickets", params=params)

    def get(self, ticket_id: str) -> Dict[str, Any]:
        """Get support ticket detail"""
        return self._client.request("GET", f"/shipment/support/tickets/{ticket_id}")

    def add_message(self, ticket_id: str, data: Dict[str, Any]) -> Dict[str, Any]:
        """Add a message to a ticket"""
        return self._client.request("POST", f"/shipment/support/tickets/{ticket_id}/messages", json=data)

File: rust/src/support_ticket.rs
use reqwest::Method;

use crate::error::Result;
use crate::http_client::HttpClient;
use crate::types::{PaginatedResponse, SupportTicket, TicketMessage};

pub struct SupportTicketClient {
    pub(crate) inner: HttpClient,
}

impl SupportTicketClient {
    pub fn new(inner: HttpClient) -> Self {
        Self { inner }
    }
}

impl SupportTicketClient {
    pub async fn list(
        &self,
        page: i32,
        page_size: i32,
        status: Option<&str>,
    ) -> Result<PaginatedResponse<SupportTicket>> {
        let mut q = format!("?page={}&page_size={}", page, page_size);
        if let Some(s) = status {
            q.push_str(&format!("&status={}", urlencoding::encode(s)));
        }
        self.inner
            .request(
                &format!("/api/v1/support/tickets{}", q),
                Method::GET,
                None,
            )
            .await
    }

    pub async fn get(&self, id: &str) -> Result<SupportTicket> {
        self.inner
            .request(
                &format!("/api/v1/support/tickets/{}", id),
                Method::GET,
                None,
            )
            .await
    }

    pub async fn create(
        &self,
        body: serde_json::Value,
    ) -> Result<SupportTicket> {
        self.inner
            .request(
                "/api/v1/support/tickets",
                Method::POST,
                Some(body),
            )
            .await
    }

    pub async fn update(
        &self,
        id: &str,
        body: serde_json::Value,
    ) -> Result<SupportTicket> {
        self.inner
            .request(
                &format!("/api/v1/support/tickets/{}", id),
                Method::PUT,
                Some(body),
            )
            .await
    }

    pub async fn add_message(
        &self,
        ticket_id: &str,
        body: serde_json::Value,
    ) -> Result<TicketMessage> {
        self.inner
            .request(
                &format!("/api/v1/support/tickets/{}/messages", ticket_id),
                Method::POST,
                Some(body),
            )
            .await
    }

    pub async fn get_messages(
        &self,
        ticket_id: &str,
        page: i32,
        page_size: i32,
    ) -> Result<PaginatedResponse<TicketMessage>> {
        self.inner
            .request(
                &format!(
                    "/api/v1/support/tickets/{}/messages?page={}&page_size={}",
                    ticket_id, page, page_size
                ),
                Method::GET,
                None,
            )
            .await
    }
}
